packages/core-app-api/src/lib/AuthConnector/DefaultAuthConnector.ts

```diff
@@ -77,12 +81,37 @@ export class DefaultAuthConnector<AuthSession>
       provider,
       joinScopes = defaultJoinScopes,
       oauthRequestApi,
+      usePopup,
       sessionTransform = id => id,
     } = options;
 
     this.authRequester = oauthRequestApi.createAuthRequester({
       provider,
-      onAuthRequest: scopes => this.showPopup(scopes),
+      onAuthRequest: async scopes => {
+        if (!usePopup) {
+          // modal before redirect
+          const scope = this.joinScopesFunc(scopes);
+          const redirectUrl = await this.buildUrl('/start', {
+            scope,
+            origin: window.location.origin,
+            redirectUrl: window.location.href,
+            authType: 'redirect',
+          });
+
+          if (provider.hasOwnProperty('provider_id')) {
+            // set the sign in provider here
+            localStorage.setItem(
+              '@backstage/core:SignInPage:provider',
+              provider.provider_id!,
+            );
+          }
```

Comment:
This is leaking pretty far across abstractions tbh. From what I can see the problem is that we're not setting the selected auth provider in the `MultiSignInPage` until auth is successful, but we redirect before that happen. That in turn means the user has to pick the same auth provider again once redirect auth is complete.

It's a pretty tricky refactor though. We could communicate it through query params, but again we'd be stuck with the ID we need being specific to the `MultiSignInPage`. An optimal solution would be that is completely housed within the `useSignInProviders` hook that the page uses. The components used there currently have a `onSignInSuccess` prop that's used to set the value in local storage. Do you think it would work if we flipped it around so that the local storage value is set as soon as the user selects one of the sign-in providers, but we clear it if sign-in fails, using a new `onSignInFailure` prop? Honestly it might just work to set it as early too though, not bothering with clearing it, I think the experience for the user might be the same.

Reply:
"That in turn means the user has to pick the same auth provider again once redirect auth is complete."

By setting the auth provider in local storage before the redirect, when the page is reloaded, the previously selected auth provider is automatically executed. So I believe that we have already implemented most of the proposed solution, but maybe the local storage property is not cleared if sign-in fails. Please let me know if I am not understanding your comment correctly.

Reply:
@headphonejames I was explaining the broken behavior that this code fixes, but at the same time we can't really have this code live here. It's too far separated from where this local storage value is actually being used and breaks the abstraction.

Coming back to this I think my last line is likely the way to go. Basically we change the sign-in page to set the selected sign-in provider in local storage as soon as you click the button, rather than in `onSignInSuccess`. The only downside of that is that if the user doesn't go through with the sign-in then it'll cause a request to the `/refresh` endpoint for that provider on subsequent loads, but I think that's fine.

Reply:
Please let me know if you prefer an onSignInFailure prop, instead of just removing the PROVIDER_STORAGE_KEY from localstorage if login canceled or threw an error.